# Chi-Square Tests

***

In [6]:
# Data frames.
import pandas as pd

# Shuffles.
import random

# Statistics.
import scipy.stats as ss

## Laerd Statistics Chi-Square Test for Independence

https://statistics.laerd.com/spss-tutorials/chi-square-test-for-association-using-spss-statistics.php

***

In [7]:
# Sixteen Males who preferred Books.
male_books = [['Male', 'Books']] * 16

# Show.
male_books

[['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books']]

In [8]:
# Sixteen Females who preferred Books.
female_books = [['Female', 'Books']] * 13

# Show.
female_books

[['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books']]

In [9]:
# Twenty-four Males who preferred Online.
male_online = [['Male', 'Online']] * 24

# Show.
male_online

[['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online']]

In [10]:
# Twenty-seven Females who preferred Online.
female_online = [['Female', 'Online']] * 27

# Show.
female_online

[['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online']]

In [11]:
# Raw data, merge the four lists.
raw_data = male_books + male_online + female_books + female_online

# Show.
raw_data

[['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Books'],


In [12]:
# Shuffle the data.
random.shuffle(raw_data)

# Show.
raw_data

[['Female', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Books'],
 ['Female', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Female', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Online'],
 ['Female', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Online'],
 ['Female', 'Books'],
 ['Male', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Male', 'Books'],
 ['Male', 'Books'],
 ['Female', 'Books'],
 ['Female', 'Online'],
 ['Female', 'Books'],
 ['Male', 'Online'],
 ['Female', 'Online'],
 ['Fem

In [13]:
# Zip the list - make the rows columns and the columns rows.
# Interchanges the outer and inner lists.
gender, medium = list(zip(*raw_data))

# Show.
gender, medium

(('Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Male',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Female',
  'Female',
  'Male',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Male',
  'Male',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Female',
  'Male',
  'Male',
  'Female',
  'Female',
  'Male',
  'Female',
  'Male',
  'Female',
  'Male',
  'Male'),
 ('Books',
  'Online',
  'Online',
  'Books',
  'Books',
  'Online',
  'Books',
  'Online',
  'Books',
  'Online',
  'B

In [14]:
# Create a data frame.
df = pd.DataFrame({'gender': gender, 'medium': medium})

# Show.
df

gender  medium
0   Female   Books
1   Female  Online
2   Female  Online
3     Male   Books
4   Female   Books
..     ...     ...
75  Female   Books
76    Male  Online
77  Female  Online
78    Male   Books
79    Male  Online

[80 rows x 2 columns]

## Contingency Table

***

In [15]:
# It does not give us what we want, just the different observations and their count
res = ss.contingency.crosstab(df)
res

CrosstabResult(elements=(array(['Books', 'Female', 'Male', 'Online'], dtype=object),), count=array([29, 40, 40, 51]))

In [16]:
# Perform Crosstabs Contingency.
# Returns a table of counts for each possible unique combination in *args
# areguments of sequences must be the same length
cross = ss.contingency.crosstab(df['gender'], df['medium'])

# Show. (13 females prefer books, 27 prefer online; 16 males prefer books, 24 prefer online)
cross

CrosstabResult(elements=(array(['Female', 'Male'], dtype=object), array(['Books', 'Online'], dtype=object)), count=array([[13, 27],
       [16, 24]]))

In [17]:
# The first variable values, and the second.
first, second = cross.elements

# Show.
first, second

(array(['Female', 'Male'], dtype=object),
 array(['Books', 'Online'], dtype=object))

In [18]:
# The counts.
cross.count

array([[13, 27],
       [16, 24]])

In [19]:
# Find all rows in df with gender equal to the first value in first.
df[df['gender'] == first[0]]

gender  medium
0   Female   Books
1   Female  Online
2   Female  Online
4   Female   Books
10  Female   Books
11  Female  Online
12  Female  Online
13  Female  Online
14  Female  Online
18  Female   Books
19  Female  Online
20  Female  Online
21  Female   Books
24  Female  Online
27  Female  Online
30  Female  Online
32  Female  Online
34  Female   Books
38  Female  Online
41  Female   Books
42  Female  Online
43  Female   Books
45  Female  Online
46  Female  Online
47  Female  Online
48  Female  Online
50  Female   Books
51  Female   Books
52  Female  Online
54  Female  Online
55  Female   Books
59  Female  Online
64  Female  Online
67  Female  Online
68  Female  Online
69  Female  Online
72  Female   Books
73  Female  Online
75  Female   Books
77  Female  Online

In [20]:
# Find all rows in df with gender equal to the first value in first and medium equal to first value in second.
df[(df['gender'] == first[0]) & (df['medium'] == second[0])]

gender medium
0   Female  Books
4   Female  Books
10  Female  Books
18  Female  Books
21  Female  Books
34  Female  Books
41  Female  Books
43  Female  Books
50  Female  Books
51  Female  Books
55  Female  Books
72  Female  Books
75  Female  Books

## Statistical Test

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html

***

In [21]:
# The contigency table.
cross.count

array([[13, 27],
       [16, 24]])

In [22]:
# Do the statistics. Just do them
# Second row in Laerd Statistics (Continuity Correction)
result = ss.chi2_contingency(cross.count, correction=False)

# Show.
result

Chi2ContingencyResult(statistic=0.486815415821501, pvalue=0.4853513240525321, dof=1, expected_freq=array([[14.5, 25.5],
       [14.5, 25.5]]))

In [23]:
# The expected fequencies if independent.
result.expected_freq

array([[14.5, 25.5],
       [14.5, 25.5]])

In [24]:
# Preferred books irrespective of gender.
29 / 80

0.3625

In [25]:
# If no relationship between gender and medium,
# then we should have same proportion of males
# liking books as we have overall.
40 * (29 / 80)

14.5

In [26]:
# If no relationship between gender and medium,
# then we should have same proportion of people liking
# online who are female as we have overall.
51 * (40 / 80)

25.5

## Bonus: Calculating the Statistic

https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test#Calculating_the_test-statistic

***

Degrees of Freedom
If I know the row and column totals:
40 each for gender (rows)
29 for books (column 1)
51 for online (column 2)
The degree of freedom is 1, because there is only 1 unknown. If I know 1 value I can calculate all of the other 3 values.

In [27]:
# Observed.
cross.count

array([[13, 27],
       [16, 24]])

In [28]:
# Expected Frequency
# 14.5 females should prefer books and 25.5 females should prefer online (the same for males)
# Because a total of 29 prefer books and 51 prefer online. If no relationship between gender 
# and medium you should get the same numbers for male and female. Hence the expected frequency.
result.expected_freq

array([[14.5, 25.5],
       [14.5, 25.5]])

The null hypothese is that there is no relationship between gender and medium.

The lower the p-value the less chance of them having a relationship.
A p-value of 0.485 tells us that there is not enough evidence to reject the null hypoothesis.

In [29]:
# Preferred books irrespective of gender
29/80

0.3625

In [30]:
# If not relationship between gender and medium we should get the same proportions for male and
# female (14.5)
# Books
40 * 29/80

14.5

In [31]:
#If not relationship between gender and medium we should get the same proportions for male and 
# female (25.5)
# Online
40 * 51/80

25.5

In [32]:
# Difference.
cross.count - result.expected_freq

array([[-1.5,  1.5],
       [ 1.5, -1.5]])

In [33]:
# Difference squared.
(cross.count - result.expected_freq)**2

array([[2.25, 2.25],
       [2.25, 2.25]])

In [34]:
# Difference squared divided by expected.
(cross.count - result.expected_freq)**2 / result.expected_freq

array([[0.15517241, 0.08823529],
       [0.15517241, 0.08823529]])

In [35]:
# Sum of (difference squared divided by expected).
((cross.count - result.expected_freq)**2 / result.expected_freq).sum()

0.486815415821501

***

## End